### PR TITLE
Fix SwapMasterPoint

### PR DESCRIPTION
**Pull request by markuspf**
If both bags were not on any list (the link pointers pointed to their
respective master pointers), then `SwapMasterPoint` just messed them up.

Unfortunately I have not been able to trigger the wrong behaviour in a test yet (short of using `SWITCH_OBJ` in `ConvertToVectorRep` and running `tst/testbugfix/2016-03-01-t00331.tst`, maybe @ChrisJefferson, @fingolfin, or @rbehrends  have an idea what exactly one needs to do to make that function misbehave in the original version.

Note also that this "fix" might not be correct yet. The `SwapMasterPoint` function has not been used or tested at all, judging from its history.

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
I want to just ponder this for a while, but I'm basically happy with it.

It's dumping a small optimisation -- before we tried to avoid marking bags as changed when we swapped master pointers.

I can't figure out why that optimisation would break things, but it does "feel" dodgy, the bags were changing in some sense.

**Comment by fingolfin** [comment]
Can you explain why exactly this helps, resp. in which way it messed up things before?

**Comment by markuspf** [comment]
Also note that `SWITCH_OBJ` actually then calls `CHANGED_BAG` on `obj1` and `obj2` after calling `SwapMasterPoint`

**Comment by markuspf** [comment]
So to help tracking this down: To trigger the wrong behaviour use #1632 and run
```
gap> g := PSL(6,4);;
gap> ConjugacyClasses(g);
```
which will (probably) crash.

**Comment by ChrisJefferson** [comment]
OK, now I understand the problem. This text might want to go in as a comment, for future viewers :)

`LINK_BAG` is used to connect up bags which have been changed since the last garbage collection. We assume that `LINK_BAG(x) == x` means the bag hasn't changed.

However, young bags (which is where `PTR_BAG(bag) > YoungBags`) are assumed to always be changed, so we never both changing `LINK_BAG` for such bags.

This means if we swap a young bag and an old bag, we have to always make sure we mark the old bag has changed, as this bag was created since the last GC so it has changed (or at least, we have no way of knowing if it needs to be checked by GC).

This raises another question -- should `SwapMasterPointer` be careful about not fiddling the `LINK_BAG` of young bags? I think gasman is happy with young bags ending up with `LINK_BAG`, but it still might be worth carefully checking.

File: src/gasman.c [hunk 1]
```diff
@@ -2082,56 +2082,43 @@ void CheckMasterPointers( void )
...
-    Bag swapbag;
+    Bag * ptr1;
+    Bag * ptr2;
+    Bag   swapbag;
 
-    if ( bag1 == bag2 )
+    if (bag1 == bag2)
         return;
 
-    /* get the pointers                                                    */
```
Comment (ChrisJefferson) [inline src/gasman.c]: Here is a suggested mega-comment :)

```
    // We need to make sure the correct things are GCed. There are various things
    // to consider.
    //
    // As changed bags are stored as a singly linked list, we can't easily
    // take anything out of the list of changed bags. Also, as the changed bag are
    // stored in a linked list with the pointer stored inside the bag itself,
    // when we swap PTR_BAG, we just to swap LINK_BAG as well.
    //
    // If both bags are old, if the original bag1 was changed we need to
    // now mark bag2 changed (and vice-versa).
    //
    // If both bags are new, GAP assumes they are both changed.
    //
    // If bag1 was young and bag2 was old, then after swap bag1 is old and
    // bag2 is young (the 'young'ness moves with the contents)
    //
    // Mark bag1 changed if bag2 was previously changed (for same reason as above).
    //
    // More importantly, bag2 is now young, but the only reference to bag2
    // might be in an old, unchanged bag, so it will get GCed.
    //
    // We therefore force bag2 onto the Changed bag list (we can't use CHANGED_BAG,
    // as it skips young bags). Gasman treats anything on the changed bag list as
    // a root, not garbage collecting it or anything it references.
    //
    // In practice, just always put both bags on the Changed bag list, rather than
    // pick through the exact cases (it is never incorrect to mark something
    // changed.)
```

**Comment by markuspf** [comment]
Updated the patch a bit, and rebased it on current master. I added @ChrisJefferson's comment verbatim. Now going to read it.

File: src/gasman.c [hunk 2]
```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
...
-    Bag                 bag2 )
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are GCed. There are various things to
+// consider.
+//
+// As changed bags are stored as a singly linked list, we can't easily take
+// anything out of the list of changed bags. Also, as the changed bags are stored
+// in a linked list with the pointer stored inside the bag itself, when we swap
+// PTR_BAG, we swap LINK_BAG as well.
```
Comment (fingolfin) [inline src/gasman.c]: perhaps "we *must* swap" ?

File: src/gasman.c [hunk 3]
```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
...
+//
+// If both bags are old, if the original bag1 was changed we need to now mark
```
Comment (fingolfin) [inline src/gasman.c]: "was changed" -> "was in the list of changed bags" ?

File: src/gasman.c [hunk 4]
```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
...
+// bag2 changed (and vice-versa).
+//
+// If both bags are young, GAP assumes they are both changed.
```
Comment (fingolfin) [inline src/gasman.c]: I think this is quite confusing to read... what does it mean that "both bags are changed" ? I think that's not the relevant thing we need to convey here. Indeed, it is not quite clear to me *what* the message here is. Perhaps: "If both bags are young, there is no need to either of them into the list of changed bags" ? But that would be kind of redundant or at least not helpful information (at least at this point), so I'd omit it.

But perhaps you meant something else?

File: src/gasman.c [hunk 5]
```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
...
+//
+// If bag1 was young and bag2 was old, then after swapping, bag1 is old and bag2 is
+// young (the 'young'ness moves with the contents)
```
Comment (fingolfin) [inline src/gasman.c]: Add trailing period.

File: src/gasman.c [hunk 6]
```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
...
+//
+// Mark bag1 changed if bag2 was previously changed (for same reason as above).
+//
+// More importantly, bag2 is now young, but the only reference to bag2 might be
+// in an old, unchanged bag, so it would get GCed.
```
Comment (fingolfin) [inline src/gasman.c]: Why singular? "... the only references to bag2 might be in old bags not on the list of changed bags. Hence bag2 would be (incorrectly!) garbage collected"

File: src/gasman.c [hunk 7]
```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
...
+//
+// We therefore force bag2 onto the ChangedBags list (we can't use CHANGED_BAG,
+// as it skips young bags). Gasman treats anything on the ChangedBags list as a
+// root, not garbage collecting it or anything it references.
```
Comment (fingolfin) [inline src/gasman.c]: Actually that is not quite correct: GASMAN treats every old bag on the changed list as a root, but young bags get special treatment: They are not treated as roots, but nevertheless marked and thus not collected.

"Gasman never garbage collects any bags on the list of changed bag"

File: src/gasman.c [hunk 8]
```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
...
+//
+// In practice, we just always put both bags on the ChangedBags list, rather than
+// pick through the exact cases (it is never incorrect to mark something changed.)
```
Comment (fingolfin) [inline src/gasman.c]: Hmm, perhaps start with this (i.e. what we do), and then explain it?

**Comment by markuspf** [comment]
Updated the comment & the commit message. Tests passed on my repository's travis, but got stuck here. I blame travis.

File: src/gasman.c [hunk 9]
```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
...
+// exact cases, as it is never incorrect to mark something changed.
+//
+// For completeness and future reference here are the necessary points to
+// consider.
+//
+// As ChangedBags is a singly linked list, we can't easily take anything out of
+// this list. Also, as the link pointers are stored inside the bag itself, when
+// we swap PTR_BAG, we must swap LINK_BAG as well.
+//
+// If both bags are old, and the original bag1 was changed, we need to now mark
```
Comment (fingolfin) [inline src/gasman.c]: I'd still change "bag1 was changed" to e.g. "bag1 had been marked as changed" or "bag1 was on the changed list" or so

File: src/gasman.c [hunk 10]
```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
...
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are garbage collected, so we always put
+// *both* bags on the ChangedBags linked-list, rather than pick through the
+// exact cases, as it is never incorrect to mark something changed.
+//
+// For completeness and future reference here are the necessary points to
+// consider.
+//
+// As ChangedBags is a singly linked list, we can't easily take anything out of
```
Comment (fingolfin) [inline src/gasman.c]: At this point, reader of this comment have no idea why they should care about removing something from the ChangedBags list. So I think this whole explanation is a bit backwards.

File: src/gasman.c [hunk 11]
```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
...
+// this list. Also, as the link pointers are stored inside the bag itself, when
+// we swap PTR_BAG, we must swap LINK_BAG as well.
+//
+// If both bags are old, and the original bag1 was changed, we need to now mark
+// bag2 changed (and vice-versa).
+//
+// If both bags are young, GAP assumes they are both changed.
```
Comment (fingolfin) [inline src/gasman.c]: It seems my previous comments are now all hidden, due to the revised commit, yet they were not all addressed :-(.

In particular, I still don't understand what the purpose of this sentence is.... Anyway, all in all, how about this after "For completeness and future reference ...":

"When swapping two master pointers, we have to take into account whether they are marked as changed or not, as we otherwise may end up in an inconsistent state, were some object is referenced by another, but GASMAN is lead to believe this is not so; GASMAN then collects the bag, resulting in a corrupted workspace.

To understand this better, we consider the following three cases:

1. Both bags are old. Then if the original bag1 had been previously marked as changed (by having been put into the ChangedBags simply linked list), then we must make sure to mark the new bag2 as changed, too (and vice-versa).

2. Both bags are young. Then they typically will not be on the list of changed bags (GAP implicitly treats all young bags as if they had been marked changed), so nothing needs to be done. (However, while CHANGED_BAG will indeed never put a young bag on the list of changed bags, this can still happen in step 3, so strictly speaking we need to do something similar in step 2 as in step 1).

3. Bag1 is young and bag2 is old (or vice-versa). Then [...]
"
now continue with the existing comment.

File: src/gasman.c [hunk 12]
```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
...
+//
+// If bag1 was young and bag2 was old, then after swapping, bag1 is old and bag2
+// is young, as the 'young'ness moves with the contents, so we must mark bag1
+// changed if bag2 was previously changed.
+//
+// More importantly, bag2 is now young, but the only references to bag2 might be
+// in an old, unchanged bag, so it would get (incorrectly) garbage collected,
```
Comment (fingolfin) [inline src/gasman.c]: perhaps change "garbage collected" to just "collected?

File: src/gasman.c [hunk 13]
```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
...
+// because these bags are not considered in a garbage collection.
+//
+// We therefore force bag2 onto the ChangedBags list, but we can't use
```
Comment (fingolfin) [inline src/gasman.c]: "We therefore" makes it sound as if this is an obvious consequence, but to me it isn't, resp. to understand it I actually need the *next* paragraph. So how about this:

"To avoid this, we for bag2 onto the ChangedBags list (note that we cannot use CHANGED_BAG for this, as it skips young bags). This prevents bag2 from being collected, as GASMAN never collects any bags on the ChangedBag list."

Then the next paragraph can be removed, too.

**Comment by fingolfin** [comment]
After having looked at GASMAN, and some discussion with @ChrisJefferson on the way to dinner, I *thought* for a while we'd also need one more change:
```patch
--- a/src/gasman.c
+++ b/src/gasman.c
@@ -1713,10 +1713,7 @@ again:
         first = ChangedBags;
         ChangedBags = LINK_BAG(first);
         LINK_BAG(first) = first;
-        if ( PTR_BAG(first) <= YoungBags )
-            (*TabMarkFuncBags[TNUM_BAG(first)])( first );
-        else
-            MarkBag(first);
+        (*TabMarkFuncBags[TNUM_BAG(first)])( first );
     }
```
Looking at the code, we treat young bags on the `ChangedBags` list differently (which is remarkable, because young bags normally can't end up on there -- the only way I know is to swap master pointers. But the code for that was only added 1997-01-10 by Frank Celler, while that special case was already in the initial CVS import from 1996-07-03... So why was it there? We might never know...

Anyway, back to topic: I thought the above was wrong, because now the "new" young bag is marked, but its subbags aren't, which could lead to trouble (i.e. those subbags incorrectly being collected).

*However*, this is not actually true, because the collection code actually continues like this:
```C
    while ( MarkedBags != 0 ) {
        first = MarkedBags;
        MarkedBags = LINK_BAG(first);
        LINK_BAG(first) = MARKED_ALIVE(first);
        (*TabMarkFuncBags[TNUM_BAG(first)])( first );
        nrLiveBags++;
        sizeLiveBags += SIZE_BAG(first);
    }
```
So, we now apply recursion to the marking. So I think now that all is fine -- but we might still apply the above patch. Or even its reverse:
```patch
--- a/src/gasman.c
+++ b/src/gasman.c
@@ -1713,10 +1713,7 @@ again:
         first = ChangedBags;
         ChangedBags = LINK_BAG(first);
         LINK_BAG(first) = first;
-        if ( PTR_BAG(first) <= YoungBags )
-            (*TabMarkFuncBags[TNUM_BAG(first)])( first );
-        else
-            MarkBag(first);
+        MarkBag(first);
     }
```

**Comment by fingolfin** [comment]
To build on my previous comment, I now think that actually both patches I included are wrong, and the current GASMAN code is "right" (at least for our purposes):
* doing `(*TabMarkFuncBags[TNUM_BAG(first)])( first );` puts the *subbags* on `MarkedBags` list, but not the bag itself
* calling `MarkBag(first)` puts just the bag (not its subbags) into the `MarkedBags` list.

Then later, the loop over `MarkedBags` marks all bags on it as alive. So only if we use `MarkBag(first)` on our swapped young bag do we get the desired effect that it stays alive during the garbage collection. Hence my first patch is "wrong", at least for our purposes)

On the other hand, `MarkBag` only does something for *young* bags. Hence the second patch is correct.

In summary, we should add a comment to that `if/else` which explains these subtleties.

File: src/gasman.c [hunk 14]
```diff
@@ -2081,54 +2081,75 @@ void CheckMasterPointers( void )
...
-*F  SwapMasterPoint( <bag1>, <bag2> ) . . . swap pointer of <bag1> and <bag2>
-*/
-void SwapMasterPoint (
-    Bag                 bag1,
-    Bag                 bag2 )
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are garbage collected, so we always put
+// *both* bags on the ChangedBags linked-list, rather than pick through the
+// exact cases, as it is never incorrect to mark something changed.
```
Comment (fingolfin) [inline src/gasman.c]: Please don't hit me, but, despite *me* having suggested this wording, I now wonder whether using the term "mark" here might be confusing, because there is also a list of "marked" bugs, and we "mark" bags as part of the mark-and-sweep. So somebody might misinterpret "mark" here? But maybe I am overly nervous here?

An alternative wording here (and later on) might be: "..never incorrect to put a bag into the ChangedBags list" ?

File: src/gasman.c [hunk 15]
```diff
@@ -2081,54 +2081,75 @@ void CheckMasterPointers( void )
...
+//
+// For completeness and future reference here are the necessary points to
+// consider.
+//
+// When swapping two master pointers, we have to take into account whether the
+// bags they refer to are marked as changed or not, as we otherwise may end up
+// in an inconsistent state, were a bag is referenced, but GASMAN does not know
```
Comment (fingolfin) [inline src/gasman.c]: were -> where?

File: src/gasman.c [hunk 16]
```diff
@@ -2081,54 +2081,75 @@ void CheckMasterPointers( void )
...
+//
+// We consider the following three cases:
+//
+// 1. Both bags are old. Then if the original bag1 had been previously marked as
+//    changed (by having been put into the ChangedBags singly linked list), then
+//    we must make sure to mark the new bag2 as changed, too (and vice-versa).
+//
+// 2. Both bags are young. Then they typically will not be on the list of
+//    changed bags, as GAP implicitly treats all young bags as if they had been
+//    marked changed.
```
Comment (fingolfin) [inline src/gasman.c]: Actually, I think now this sentence (which is of course mine, so all blame to me) is wrong or at least misleading. Perhaps just say: "... changed bags, as CHANGED_BAGS ignores young bags. However, young bags can still [... continue as before ]"

File: src/gasman.c (hunk 14 above)
Comment (markuspf) [reply src/gasman.c]: I as wondering the same, hence tried to use the "being on the `ChangedBags`" formulation, albeit not consistently.

File: src/gasman.c (hunk 16 above)
Comment (markuspf) [reply src/gasman.c]: In your defense, I rephrased some of the comments a bit, so this might be entirely my fault.
